### PR TITLE
FINERACT-1847: Removing Edit function from Goodwill Credit or Payout Refund or Merch…

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

Fix on API level which will remove the Edit option of `Goodwill Credit`, `Payout Refund` and `Merchant Issued Refund` transactions

[FINERACT-1847](https://issues.apache.org/jira/browse/FINERACT-1847)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Please double check the tests. It is failing with this:

```
java.lang.NullPointerException: Cannot invoke "com.google.gson.JsonElement.getAsJsonArray()" because the return value of "com.google.gson.JsonObject.get(String)" is null
[2138](https://github.com/apache/fineract/actions/runs/3825276813/jobs/6508131758#step:11:2139)
      at org.apache.fineract.integrationtests.BatchApiTest.shouldReturnOkStatusForBatchGoodwillCreditReversal(BatchApiTest.java:1815)
```

**Comment by josehernandezfintecheandomx** [comment]
> @josehernandezfintecheandomx Please double check the tests. It is failing with this:
> 
> ```
> java.lang.NullPointerException: Cannot invoke "com.google.gson.JsonElement.getAsJsonArray()" because the return value of "com.google.gson.JsonObject.get(String)" is null
> [2138](https://github.com/apache/fineract/actions/runs/3825276813/jobs/6508131758#step:11:2139)
>       at org.apache.fineract.integrationtests.BatchApiTest.shouldReturnOkStatusForBatchGoodwillCreditReversal(BatchApiTest.java:1815)
> ```

Done! @adamsaghy I needed to update this test because the reversal or edit for these transactions are not allowed more

**Comment by galovics** [comment]
@josehernandezfintecheandomx test is still failing

```

  Test applyLoanTransactionChargeback() FAILED

  java.lang.AssertionError: 1 expectation failed.
  Expected status code <403> but was <503>.
      at org.apache.fineract.integrationtests.LoanTransactionChargebackTest.applyLoanTransactionChargeback(LoanTransactionChargebackTest.java:120)
```

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -1063,9 +1063,10 @@ public CommandProcessingResult adjustLoanTransaction(final Long loanId, final Lo
                     transactionId);
         }
 
-        if (transactionToAdjust.isChargeback()) {
-            throw new PlatformServiceUnavailableException("error.msg.loan.transaction.update.not.allowed", "Loan transaction:"
-                    + transactionId + " update not allowed as loan transaction is charge back and is linked to other transaction",
+        if (transactionToAdjust.isChargeback() || transactionToAdjust.isGoodwillCredit() || transactionToAdjust.isPayoutRefund()
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: This is not correct. We should let the undo, but the adjusting.
Please amend this logic whether it is just undo (transactionAmount=0) or undo and create new (transactionAmount>0)

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/BatchApiTest.java [hunk 2]
```diff
@@ -1808,19 +1806,16 @@ public void shouldReturnOkStatusForBatchGoodwillCreditReversal() {
 
         final List<BatchResponse> responses = BatchHelper.postBatchRequestsWithoutEnclosingTransaction(this.requestSpec, this.responseSpec,
                 BatchHelper.toJsonString(batchRequests));
+        Assertions.assertEquals(HttpStatus.SC_INTERNAL_SERVER_ERROR, (long) responses.get(4).getStatusCode(),
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/BatchApiTest.java]: Undo goodwill credit is supported. Please revert this test case.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/BatchApiTest.java [hunk 3]
```diff
@@ -1830,7 +1825,7 @@ public void shouldReturnOkStatusForBatchGoodwillCreditReversal() {
      * @see AdjustTransactionCommandStrategy
      */
     @Test
-    public void shouldReturnOkStatusOnSuccessfulTransactionMerchantIssuedAndPayoutRefundReversal() {
+    public void shouldReturnNOkStatusOnUnSuccessfulTransactionMerchantIssuedAndPayoutRefundReversal() {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/BatchApiTest.java]: Undo merchant issues refund is supported. Please revert this test case.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java [hunk 4]
```diff
@@ -438,4 +439,76 @@ public void newGoodwillCreditCreatesCorrectJournalEntriesForCashAccountingTest()
 
     }
 
+    @Test
+    public void tryToUndoGoodWillCreditTransactionTest() {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java]: Please amend this test cases to tests for testing whether:
- these transactions can be undone
- these transactions cannot be adjusted

Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 5]
```diff
@@ -913,6 +913,15 @@ public CommandWrapperBuilder adjustTransaction(final Long loanId, final Long tra
         return this;
     }
 
+    public CommandWrapperBuilder undoTransaction(final Long loanId, final Long transactionId) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java]: Dead code?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please add test cases to cover the adjusting of the goodwill credit, payout refund and merchant issued refund and chargeback is not possible and they throw the proper error.

Also please remove the "dead code" which was added (CommandWrapperBuilder.java)

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java [hunk 6]
```diff
@@ -114,7 +116,7 @@ public void applyLoanTransactionChargeback() {
 
         // Try to reverse a Loan Transaction charge back
         PostLoansLoanIdTransactionsResponse reverseTransactionResponse = loanTransactionHelper.reverseLoanTransaction(loanId,
-                chargebackTransactionId, operationDate, responseSpecErr503);
+                chargebackTransactionId, operationDate, responseSpecErr403);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java:119]: So reversing or adjusting a chargeback is not possible, right?

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java]: Done!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionChargebackTest.java:119]: Yes, I needed to change the HTTP code due I sync the Exception used in this case to `InvalidLoanTransactionTypeException`

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java [hunk 7]
```diff
@@ -438,4 +439,73 @@ public void newGoodwillCreditCreatesCorrectJournalEntriesForCashAccountingTest()
 
     }
 
+    @Test
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java:444]: Please add test cases to cover the adjusting of the goodwill credit, payout refund and merchant issued refund and chargeback is not possible and they throw the proper error.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Missing test cases:
Please add test cases to cover the adjusting of the goodwill credit, payout refund and merchant issued refund and chargeback is not possible and they throw the proper error.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest.java:444]: The IT were added in the class `ClientLoanCreditBalanceRefundandRepaymentTypeIntegrationTest` 

Screenshot 2023-01-11 at 14 32 40

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@josehernandezfintecheandomx Testing whether adjusting 
- merchant issued refund 
- chargeback

is not enabled is still missing. Would you please add the test cases for these two as well?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java [hunk 8]
```diff
@@ -430,7 +430,8 @@ public CommandProcessingResult updateClient(final Long clientId, final JsonComma
             }
 
             final ExternalId externalId = externalIdFactory.createFromCommand(command, ClientApiConstants.externalIdParamName);
-            if (command.isChangeInStringParameterNamed(ClientApiConstants.externalIdParamName, externalId.getValue())) {
+            if (command.isChangeInStringParameterNamed(ClientApiConstants.externalIdParamName,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java]: Please raise a separate PR for fixing the client external id updating bug!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please extract the bugfix of "updateClient" logic into a new, separate PR!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java]: Done

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**